mx_fetchjdk.py

```diff
@@ -0,0 +1,165 @@
+#
+# ----------------------------------------------------------------------------------------------------
+#
+# Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+#
+# This code is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 2 only, as
+# published by the Free Software Foundation.
+#
+# This code is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+# or visit www.oracle.com if you need additional information or have any
+# questions.
+#
+# ----------------------------------------------------------------------------------------------------
+#
+from __future__ import print_function
+import os, shutil
+from os.path import join, exists, abspath, isdir, islink
+from shutil import copytree
+from argparse import ArgumentParser, REMAINDER
+
+from mx import optional_suite_context, _mx_home, command
+from select_jdk import find_system_jdks
+from jdk_distribution_parser import JdkDistribution
+import mx, mx_urlrewrites
+
+@command('mx', 'fetch-jdk', '[options]')
+@optional_suite_context
+def fetch_jdk(args):
+    """fetches required JDK version
+
+    If called without --quiet flag, menu will be printed for available JDK selection.
+    """
+    args = _parse_fetchjdk_settings(args)
+
+    distribution = args["java-distribution"]
+    jdk_path = args["jdk-path"]
+    jdk_folder = distribution.get_jdk_folder()
+    full_jdk_path = distribution.get_full_jdk_path(jdk_path)
+    jdk_url = mx_urlrewrites.rewriteurl(distribution.get_url())
+    jdk_archive = distribution.get_archive()
+
+    if not exists(full_jdk_path):
+        mx._opts.no_download_progress = args["quiet"]
+        mx.download(join(jdk_path, jdk_archive), [jdk_url], verbose=not args["quiet"])
+        untar = mx.TarExtractor(join(jdk_path, jdk_archive))
+
+        if not args["quiet"]:
+            print("Installing...")
+
+        with untar._open() as tar_file:
+            curr_full_jdk_path = untar._getnames(tar_file)[0]
+
+        untar.extract(jdk_path)
+        if not args["keep-archive"]:
```

Comment:
This should be in the try-catch. If something fails the archive is kept and it actually breaks the consecutive download.